### ISSUE TITLE
New operator for recursive concatenation

### ISSUE THREAD

**Issue by gabrielhuff** [ReactiveX/RxJava#5685]
How about this new operator (writing API in kotlin for readability):

```kotlin
fun <T> Single<T>.link(next: (T) -> Single<T>?) : Flowable<T>
```

**TL;DR:** It's basically a recursive `concat`.

Say you have a finite sequence of `Singles` `s1`, `s2`, `s3`, …, `sn`. The elements emitted by all singles can be concatenated by calling `concat(s1, s2, s3, ..., sn)`. However, there are scenarios in which `s2` depends on the value emitted by `s1`, `s3` depends on the value emitted by `s2` and so on.

The `link` operator would be used to concatenate sequences whose consecutive singles can be calculated as function of their predecessor's emitted value:

```
RECEIVER: --A1

OPERATOR: link { A(i) -> --A(i+1) }

OUTPUT:   --A1--A2--A3--A4--A5-|
```

Returning `null` would mean that the sequence is over. It would be equivalent to:

```kotlin
fun <T> Single<out T>.link(next: (T) -> Single<out T>?) : Flowable<T> {
    return flatMapPublisher { Flowable.concat(
            Flowable.just(it),
            next(it)?.link(next) ?: Flowable.empty()
    ) }
}
```

A typical use case would be consuming paginated data:

```kotlin
val elements = service.getFirstPage()
            .link { it.nextPage }
            .flatMap { Flowable.fromIterable(it.items) }
```

I created a single file [demo app](https://github.com/gabrielhuff/rx-pagination-demo) (the operator definition can be found [here](https://github.com/gabrielhuff/rx-pagination-demo/blob/master/src/main/kotlin/main.kt#L65-L70) and the usage [here](https://github.com/gabrielhuff/rx-pagination-demo/blob/master/src/main/kotlin/main.kt#L24))

**Comment by akarnokd** [comment]
Such operator is available in the extensions project: https://github.com/akarnokd/RxJava2Extensions/blob/master/README.md#flowabletransformersexpand

**Comment by gabrielhuff** [comment]
Cool. Any plans to include `expand` on the main lib as an operator and / or add the equivalent `Single`, `Observable` and `Maybe` implementations?

**Comment by akarnokd** [comment]
No, this is a rare operator better hosted by the extension library. Use the various `toX` to convert to the required parameter type.

**Comment by gabrielhuff** [comment]
Ok, thanks for the answer. You can close this issue now.
